### PR TITLE
feat(providers): add gemini web search provider

### PR DESCRIPTION
**Pull request by bogdanovich**
## Summary

Adds a Gemini Google Search provider for the existing `web_search` tool.

The provider calls Gemini `generateContent` with Google Search grounding and formats the synthesized answer plus grounding citations for the agent. It can be selected explicitly with `tools.web.provider = "gemini"`, and `auto` prefers the existing free providers first before falling through to Gemini and other API-key-backed providers when they are enabled and configured.

## Changes

- Add `tools.web.gemini` config with `enabled`, `api_key`, `model`, and `max_results`.
- Add `GeminiSearchProvider` and provider readiness/routing support.
- Expose Gemini in the web search config API and frontend config type.
- Add Gemini web UI fields for:
  - API key / token
  - model override
- Document the provider in README and tools configuration docs.
- Add tests for auto selection and missing-key fallback behavior.
- Add unit tests for `GeminiSearchProvider.Search()` covering:
  - normal response parsing with grounding citations
  - API error status handling
  - empty candidates
  - citation count limiting

## Validation

- `go test ./pkg/tools/integration -run 'Test(WebTool_AutoProviderPrefersFreeProvidersBeforeGemini|WebTool_GeminiRequiresAPIKey|GeminiSearchProvider_SearchSuccess|GeminiSearchProvider_SearchAPIError|GeminiSearchProvider_SearchEmptyCandidates)' -count=1`
- `go test ./pkg/config ./web/backend/api -count=1`
- Manual web UI validation on the launcher dashboard:
  - Gemini shows `Max Results`, `API Key / Token`, and `Model`
  - API key persists through the secure config path
  - model persists through `config.json`
  - explicit Gemini search works end-to-end against the live Gemini API
  - `auto` mode behavior was smoke-tested with native search disabled and Sogou disabled

## Notes

- `malformed JSON` is not a realistic manual web-UI scenario for the live Gemini endpoint; that branch is better covered by mocked/unit tests than by interactive dashboard testing.
- Gemini does not require a Base URL field in the UI.

### CODE REVIEW

**Review by SiYue-ZO — COMMENTED** [review]
Thanks for adding Gemini Google Search provider! The implementation is clean and follows the project's existing patterns well. A couple of suggestions:

**1. Provider 优先级建议调整**

当前 `auto` 模式下 Gemini 被置于最高优先级：

```go
autoPrimaryWebSearchProviders = []string{"gemini", "perplexity", "brave", "searxng", "tavily"}
```

建议将免费的 DuckDuckGo 和 Sogou 的优先级提高。理由：
- DuckDuckGo 无需 API key，无调用限制，适合作为默认兜底
- Sogou 同样免费，且对中文查询优化较好
- Gemini 需要 API key 且有调用成本，更适合用户显式选择或作为高级选项

建议调整为类似：

```go
autoPrimaryWebSearchProviders = []string{"duckduckgo", "sogou", "gemini", "perplexity", "brave", "searxng", "tavily"}
```

这样在 `auto` 模式下优先使用免费 provider，Gemini 作为需要 API key 的高级选项排在其后。

**2. 建议补充 Search 方法的测试**

当前测试覆盖了 provider 选择逻辑（自动优先级和 API key 缺失回退），但缺少 `GeminiSearchProvider.Search()` 本身的单元测试。建议使用 mock HTTP server 添加以下场景的测试：
- 正常响应解析（含 grounding citations）
- API 返回错误状态码
- 空结果（candidates 为空）
- 引用数量限制（`maxResults`）

其他部分实现简洁、安全措施到位，整体 LGTM 👍

**Comment by SiYue-ZO** [comment]
Updating my review in English:

Thanks for adding the Gemini Google Search provider! The implementation is clean and follows the project's existing patterns well. A couple of suggestions:

**1. Provider priority should be reconsidered**

Currently Gemini is placed at the highest priority in `auto` mode:

```go
autoPrimaryWebSearchProviders = []string{"gemini", "perplexity", "brave", "searxng", "tavily"}
```

I'd suggest prioritizing the free providers (DuckDuckGo and Sogou) first. Reasons:
- DuckDuckGo requires no API key and has no rate limits — ideal as the default fallback
- Sogou is also free and works well for Chinese-language queries
- Gemini requires an API key and has usage costs, making it better suited as an explicit user choice or premium option

Suggested ordering:

```go
autoPrimaryWebSearchProviders = []string{"duckduckgo", "sogou", "gemini", "perplexity", "brave", "searxng", "tavily"}
```

This way `auto` mode prefers free providers first, with Gemini (and other API-key-required providers) as fallbacks.

**2. Add tests for the Search method**

The current tests cover provider selection logic (auto priority and API key fallback), but `GeminiSearchProvider.Search()` itself has no unit tests. Consider adding tests with a mock HTTP server for:
- Normal response parsing (with grounding citations)
- API error status codes
- Empty results (no candidates)
- Citation count limiting (`maxResults`)

Everything else looks solid — clean implementation, good security practices. LGTM overall 👍

**Comment by bogdanovich** [comment]
Addressed both suggestions.

Changes:
- updated `auto` web-search resolution so the free providers are preferred before Gemini and other API-key-backed providers
- kept the existing DuckDuckGo vs Sogou query preference, so English/Latin-style queries resolve to DuckDuckGo first while Chinese-style queries resolve to Sogou first
- added unit tests for `GeminiSearchProvider.Search()` covering:
  - normal response parsing with grounding citations
  - API error status handling
  - empty candidates
  - citation count limiting

Validation:
- `go test ./pkg/tools/integration -run 'Test(WebTool_AutoProviderPrefersFreeProvidersBeforeGemini|WebTool_GeminiRequiresAPIKey|GeminiSearchProvider_SearchSuccess|GeminiSearchProvider_SearchAPIError|GeminiSearchProvider_SearchEmptyCandidates)' -count=1`
- `go test ./pkg/config ./web/backend/api -count=1`

**Review by SiYue-ZO — CHANGES_REQUESTED** [review]
### Overall Assessment: Good implementation with solid testing, but requires frontend UI completion.

### Issues Found:

**Critical: Missing Frontend UI**
- Backend fully supports Gemini configuration, but frontend lacks the input fields
- Users cannot fill in API Key from the web UI
- Location: web/frontend/src/components/agent/tools/web-search-provider-settings.tsx
- Required changes:
  1. Add gemini to apiKeyProviders set (lines 30-34)
  2. Add model input field for Gemini provider
- Workaround: Users must set via config.json or environment variables for now

**Minor: Documentation**
- Consider updating the web search tools documentation to mention Gemini's configuration requirement (API key from https://aistudio.google.com/api-keys)
- Add note that BaseURL is not required for Gemini (unlike SearXNG, GLM Search, etc.)

### Testing Recommendation:
- Verify the provider priority logic works correctly with actual Gemini API
- Test with edge cases: empty responses, malformed JSON, network errors

### Before Merging:
Please add the missing frontend UI elements so users can configure Gemini through the web interface.

**Comment by bogdanovich** [comment]
Addressed the missing frontend UI for Gemini.

Changes in this update:
- added Gemini to the web UI API-key provider set
- added a Gemini model input field in `web/frontend/src/components/agent/tools/web-search-provider-settings.tsx`

Manual validation:
- confirmed the dashboard now exposes `Max Results`, `API Key / Token`, and `Model` for Gemini
- confirmed the API key persists through the secure config path and the model persists through `config.json`
- smoke-tested live Gemini usage from the web UI with native search disabled
- also smoke-tested `auto` mode with Sogou disabled

On the testing recommendation:
- live/manual testing is appropriate for provider priority behavior and real Gemini API execution
- `malformed JSON` is not realistically triggerable from a normal Gemini query in the dashboard, so that branch is better treated as a mocked/unit-test scenario rather than a manual UI test

Also kept Gemini without a Base URL field in the UI, since the backend does not require one.

**Review by SiYue-ZO — APPROVED** [review]
@wj-xiao

**Comment by SiYue-ZO** [comment]
@bogdanovich CI didn't pass, please fix it

**Comment by bogdanovich** [comment]
@SiYue-ZO  Fixed!

**Comment by SiYue-ZO** [comment]
@wj-xiao Please take a look at this, thanks.

File: pkg/tools/integration/web.go [hunk 1]
```diff
@@ -1135,14 +1250,15 @@ var (
...
+		"gemini",
 		"brave",
 		"tavily",
 		"perplexity",
 		"searxng",
 		"glm_search",
 		"baidu_search",
 	}
-	autoPrimaryWebSearchProviders  = []string{"perplexity", "brave", "searxng", "tavily"}
+	autoPrimaryWebSearchProviders  = []string{"gemini", "perplexity", "brave", "searxng", "tavily"}
```
Comment (wj-xiao) [inline pkg/tools/integration/web.go]: This makes `auto` select Gemini before the existing free/default providers whenever a Gemini API key is configured. That changes the cost/quota behavior for users who enable Gemini, and it also contradicts the PR description saying auto should prefer free providers first. Please move Gemini after the free providers, or explicitly document and test this as an intentional behavior change.

File: pkg/tools/integration/web.go [hunk 2]
```diff
@@ -472,6 +472,113 @@ type SogouSearchProvider struct {
...
 }
 
+type GeminiSearchProvider struct {
+	apiKey string
+	model  string
+	proxy  string
+	client *http.Client
+}
+
+func (p *GeminiSearchProvider) Search(
```
Comment (wj-xiao) [inline pkg/tools/integration/web.go:482]: `GeminiSearchProvider.Search` accepts `rangeCode` but ignores it. The public `web_search` tool exposes `range`, and the other providers map it to their API-specific recency filters. With Gemini selected, users silently lose the requested time filter. Please either map the range into the Gemini prompt/request or return a clear unsupported behavior.

**Comment by bogdanovich** [comment]
Fixed the failing `Tests` check by aligning `TestWebTool_AutoProviderPrefersConfiguredProvidersBeforeGemini` with the current auto-provider priority rules. The implementation already prefers configured providers like Brave before Gemini in `auto` mode, so the failure was a stale test expectation rather than a runtime bug.

Verified locally with:
- `go test ./pkg/tools/integration -run '^TestWebTool_AutoProviderPrefersConfiguredProvidersBeforeGemini$' -count=1`
- `go test ./pkg/tools/integration -count=1`
- `go generate ./...`
- `env -u PICOCLAW_HOME go test -tags goolm,stdjson ./...`

Branch tip is `685dba1c` (`test(web): align auto-provider expectations`).

File: pkg/tools/integration/web.go [hunk 3]
```diff
@@ -472,6 +472,116 @@ type SogouSearchProvider struct {
...
+	ctx context.Context,
+	query string,
+	count int,
+	rangeCode string,
+) (string, error) {
+	if rangeCode != "" {
```
Comment (wj-xiao) [inline pkg/tools/integration/web.go]: Rejecting any non-empty `range` makes the global `web_search.range` parameter fail whenever auto mode resolves to Gemini. The tool schema advertises `range` as supported by `web_search`, so callers will not know this provider-specific limitation. Please either ignore the range for Gemini, fall back to a provider that supports recency filters when `range` is set, or make provider selection avoid Gemini for ranged searches.

**Comment by bogdanovich** [comment]
Addressed the latest range-related review feedback.

What changed:
- simplified the previous approach: Gemini no longer has resolver-level special handling for ranged searches
- removed the Gemini error for non-empty \
- Gemini now accepts the global \ parameter and ignores it, matching one of the requested options from the review
- added/updated test coverage with \

Why:
- Gemini Google Search grounding does not expose a documented structured recency/range API parameter
- ignoring the range keeps provider selection simple and avoids hardcoding Gemini capability checks into the generic resolver

Verification:
- \
- \

Pushed:
- \ — \

**Comment by bogdanovich** [comment]
Addressed the latest range-related review feedback.

What changed:
- Gemini now accepts the global `web_search.range` parameter and ignores it, matching one of the requested options from the review
- added/updated test coverage with `TestGeminiSearchProvider_SearchIgnoresRange`

Why:
- Gemini Google Search grounding does not expose a documented structured recency/range API parameter
- ignoring the range keeps provider selection simple and avoids hardcoding Gemini capability checks into the generic resolver

Verification:
- `go test ./pkg/tools/integration -run 'Test(WebTool_AutoProviderPrefersConfiguredProvidersBeforeGemini|GeminiSearchProvider_SearchIgnoresRange|WebTool_GeminiRequiresAPIKey|GeminiSearchProvider_SearchSuccess|GeminiSearchProvider_SearchAPIError|GeminiSearchProvider_SearchEmptyCandidates|WebTool_AutoProviderRoutesQueryLanguageBetweenSogouAndDuckDuckGo)' -count=1`
- `go test ./pkg/tools/integration -count=1`

Pushed:
- `3084c694` — `fix(web): let gemini ignore search range`

**Review by wj-xiao — APPROVED** [review]
